File: Ruby/excercise1/fibonacci.rb
class Fibonacci 
	def initialize(n)
		@n = n
	end
	
	def fibo1
		fiboNaive(@n)
	end

	def fibo2
		fiboReverse(@n)
	end
	
	def fibo3
	
	end
	
	private 
	
	def fiboNaive(n)
		if (n == 1 or n == 2) then 1
		else fiboNaive(n-1) + fiboNaive(n-2)
		end
	end
	
	def fiboReverse(n)
		if (n == 1 or n == 2) then 1
		else countUp(n, 2, 1, 1)
		end
	end
	
	def countUp (n, x, acc, pred)
		if x == n then acc
		else countUp(n, x + 1, acc + pred, acc)
		end
	end
end

class Matrix
	attr_accessor :matrix, :rows, :columns
	
	def initialize(m)
		@matrix = m
		@rows = m.size
		@columns = m[0].size
		if @rows != @columns then raise "not a square matrix" end
	end
	
	def multiply(other)
		if !other.is_a? Matrix then raise "argument is not a matrix" end 
		b = other.matrix
		
		result = [
					 [@matrix[0][0]*b[0][0] + @matrix[0][1] * b[1][0], @matrix[0][0]*b[0][1] + @matrix[0][1] * b[1][1]],
					 [@matrix[1][0]*b[0][0] + @matrix[1][1] * b[1][0], @matrix[1][0]*b[0][1] + @matrix[1][1] * b[1][1]]
				 ]
	  
	  Matrix.new(result)
	end
	
	def to_s
		s = ""
		@matrix.each {|m| s += m.to_s + "\n"}
		s
	end
end	

f = Fibonacci.new(20)
puts f.fibo1
puts f.fibo2

m = [[1,1], [1,0]]
matrix1 = Matrix.new(m)
matrix2 = Matrix.new(m)
print matrix1


puts matrix1.multiply(matrix2)

File: src/main/scala/excercise1/Fibonacci.scala
package excercise1

import excercise1.MatrixOps._

object Fibonacci {
  /**
   * Fibonacci number, naive implementation
   * @param n
   * @return
   */
  def fibo1(n: Int): Long = {
    if (n == 2 || n == 1) 1
    else fibo1(n - 1) + fibo1(n - 2)
  }

  /**
   * Fibonacci number, less naive implementation with
   * tail recursion and counting up instead of down.
   * @param n
   * @return
   */
  def fibo2(n: Int) = {
    def inner(x: Int, acc: BigDecimal, pred: BigDecimal): BigDecimal = {
      if (x == n) acc
      else inner(x + 1, acc + pred, acc)
    }
    if (n == 1 || n == 2) 1L
    else inner(2, 1, 1)
  }

  /**
   * Implementation using Matrices and binary exponentiation
   * @param n
   * @return
   */
  def fibo3 (n: Int): BigDecimal = {
    def inner (Z: Matrix, times: Int, y: Matrix) : Matrix = {
      if (times == 0) y
      else {
        val N = times / 2
        if (times % 2 == 0) inner(square(Z), N, y)
        else inner(square(Z), N, multiplySquare(Z, y))
      }
    }
    if (n == 1 || n == 2) 1
    else multiplyWithVector(inner(BaseMatrix, n-2, IdentityMatrix))(0)(0)
  }
}
